File: android/moqkit/moqkit/src/main/java/com/swmansion/moqkit/subscribe/internal/pipeline/TimestampDomainMapper.kt
package com.swmansion.moqkit.subscribe.internal.pipeline

import kotlin.math.abs

/**
 * Maps timestamps between active audio and video timelines without owning either timeline.
 * The timelines remain the live-edge authorities; this class only compares their domains.
 */
internal class TimestampDomainMapper(
    private val audioTimeline: () -> TrackTimeline?,
    private val videoTimeline: () -> TrackTimeline?,
) {
    fun videoOffsetUs(thresholdUs: Long): Long? {
        require(thresholdUs >= 0L) { "thresholdUs must be non-negative" }
        val (audioEdgeUs, videoEdgeUs) = liveEdgesUs() ?: return null
        val offsetUs = subtractOrNull(audioEdgeUs, videoEdgeUs) ?: return null
        return offsetUs.takeIf { absoluteValueExceeds(it, thresholdUs) }
    }

    fun audioTimeUs(videoTimeUs: Long, thresholdUs: Long): Long {
        val offsetUs = videoOffsetUs(thresholdUs) ?: return videoTimeUs
        return addOrNull(videoTimeUs, offsetUs)?.takeIf { it >= 0L } ?: videoTimeUs
    }

    fun videoTimeUs(audioTimeUs: Long, thresholdUs: Long): Long {
        val offsetUs = videoOffsetUs(thresholdUs) ?: return audioTimeUs
        return subtractOrNull(audioTimeUs, offsetUs)?.takeIf { it >= 0L } ?: audioTimeUs
    }

    fun videoTimeUsOrNull(audioTimeUs: Long, thresholdUs: Long): Long? {
        require(thresholdUs >= 0L) { "thresholdUs must be non-negative" }
        val (audioEdgeUs, videoEdgeUs) = liveEdgesUs() ?: return null
        val offsetUs = subtractOrNull(audioEdgeUs, videoEdgeUs) ?: return null
        val effectiveOffsetUs = offsetUs.takeIf { absoluteValueExceeds(it, thresholdUs) } ?: 0L
        return subtractOrNull(audioTimeUs, effectiveOffsetUs)?.takeIf { it >= 0L } ?: audioTimeUs
    }

    private fun liveEdgesUs(): Pair<Long, Long>? {
        val audioEdgeUs = audioTimeline()?.liveEdgeUs() ?: return null
        val videoEdgeUs = videoTimeline()?.liveEdgeUs() ?: return null
        return audioEdgeUs to videoEdgeUs
    }

    private fun absoluteValueExceeds(value: Long, threshold: Long): Boolean =
        value == Long.MIN_VALUE || abs(value) > threshold

    private fun subtractOrNull(left: Long, right: Long): Long? = try {
        Math.subtractExact(left, right)
    } catch (_: ArithmeticException) {
        null
    }

    private fun addOrNull(left: Long, right: Long): Long? = try {
        Math.addExact(left, right)
    } catch (_: ArithmeticException) {
        null
    }
}

File: ios/Sources/MoQKit/Subscribe/internal/pipeline/TimestampDomainMapper.swift
import Foundation

/// Maps timestamps between audio and video domains without owning either timeline.
final class TimestampDomainMapper: @unchecked Sendable {
    private let lock = UnfairLock()
    private var audioTimeline: TrackTimeline?
    private var videoTimeline: TrackTimeline?

    init(
        audioTimeline: TrackTimeline?,
        videoTimeline: TrackTimeline?
    ) {
        self.audioTimeline = audioTimeline
        self.videoTimeline = videoTimeline
    }

    func setAudioTimeline(_ timeline: TrackTimeline?) {
        lock.withLock { audioTimeline = timeline }
    }

    func setVideoTimeline(_ timeline: TrackTimeline?) {
        lock.withLock { videoTimeline = timeline }
    }

    func videoOffsetUs(thresholdUs: Int64) -> Int64? {
        precondition(thresholdUs >= 0)
        let timelines = lock.withLock { (audioTimeline, videoTimeline) }
        guard let audio = timelines.0?.liveEdgeUs(),
              let video = timelines.1?.liveEdgeUs()
        else { return nil }
        let result = audio.subtractingReportingOverflow(video)
        guard !result.overflow else { return nil }
        let offset = result.partialValue
        return Self.absoluteValue(offset, exceeds: thresholdUs) ? offset : nil
    }

    func audioTimeUs(videoTimeUs: UInt64, thresholdUs: Int64) -> UInt64 {
        guard let offset = videoOffsetUs(thresholdUs: thresholdUs),
              videoTimeUs <= UInt64(Int64.max)
        else { return videoTimeUs }
        let result = Int64(videoTimeUs).addingReportingOverflow(offset)
        guard !result.overflow, result.partialValue >= 0 else { return videoTimeUs }
        return UInt64(result.partialValue)
    }

    func videoTimeUs(audioTimeUs: UInt64, thresholdUs: Int64) -> UInt64 {
        guard let offset = videoOffsetUs(thresholdUs: thresholdUs),
              audioTimeUs <= UInt64(Int64.max)
        else { return audioTimeUs }
        let result = Int64(audioTimeUs).subtractingReportingOverflow(offset)
        guard !result.overflow, result.partialValue >= 0 else { return audioTimeUs }
        return UInt64(result.partialValue)
    }

    private static func absoluteValue(_ value: Int64, exceeds threshold: Int64) -> Bool {
        value == .min || abs(value) > threshold
    }
}
